### ISSUE TITLE
TabulatedFunction, InteractionGroups and NonbondedForce compile wrong on CUDA and OpenCL platforms

### ISSUE THREAD

**Issue by blake-armstrong** [openmm/openmm#3726]
This issue is very closely related to #1441.

I have a system where I am creating two `CustomNonbondedForce`s, with multiple tabulated `Discrete2DFunction`s in both.  When adding an `InteractionGroup` to both Force 1 and Force 2 on the OpenCL and CUDA platforms, the following errors are thrown:

- OpenCL
```
File "/home/ubuntu/blake/FEP/new_ff/case.py", line 112, in <module>
    print(context.getState(getEnergy=True).getPotentialEnergy())
  File "/home/ubuntu/miniconda3/lib/python3.9/site-packages/openmm/openmm.py", line 12318, in getState
    state = _openmm.Context_getState(self, types, enforcePeriodicBox, groups_mask)
openmm.OpenMMException: Error compiling kernel: <kernel>:220:176: error: redefinition of parameter 'table0'
        , __global const real* restrict global_nonbonded0_charge, __global const float* restrict table0, __global const float* restrict table1, __global const float* restrict table0, __global const float* restrict table1, __global const float* restrict table2) {
```
- CUDA
``` 
File "/home/ubuntu/blake/FEP/new_ff/case.py", line 112, in <module>
    print(context.getState(getEnergy=True).getPotentialEnergy())
  File "/home/ubuntu/miniconda3/lib/python3.9/site-packages/openmm/openmm.py", line 12318, in getState
    state = _openmm.Context_getState(self, types, enforcePeriodicBox, groups_mask)
openmm.OpenMMException: Error launching CUDA compiler: 256
/tmp/openmmTempKernel0x55b422fe8340_68855.cu(879): error: duplicate parameter name
/tmp/openmmTempKernel0x55b422fe8340_68855.cu(879): error: duplicate parameter name
2 errors detected in the compilation of "/tmp/tmpxft_00010d0c_00000000-6_openmmTempKernel0x55b422fe8340_68855.cpp1.ii".
```

However, no error occurs when using the CPU platform. Also, no error occurs when adding only one `InteractionGroup` to either Force 1 or Force 2, and the problem only happens when there is a `NonbondedForce` present in the system.

Here is a simple(ish) test case + particle positions, I have not tested more than `Discrete2DFunction`. I'm using the latest stable release of OpenMM installed through Conda.
```python
import openmm as mm
import openmm.app as app
import openmm.unit as unit
import numpy as np

atomTypes  = ["X" , "Y"]
numTypes = len(atomTypes)

#FORCE 1
ff_lj = []
ff_lj.append({"type1" : "Y" , "type2" : "X" , "eps" : "0.000009" , "sig" : "0.02"})
eps_r = np.zeros([numTypes,numTypes])
sig_r = np.ones([numTypes,numTypes])
for f in ff_lj:
    i = atomTypes.index(f["type1"])
    j = atomTypes.index(f["type2"])
    eps_r[i,j] = float(f["eps"])
    sig_r[i,j] = float(f["sig"])
    eps_r[j,i] = eps_r[i,j]
    sig_r[j,i] = sig_r[i,j]
epsLST_r = (eps_r).ravel().tolist()
sigLST_r = (sig_r).ravel().tolist()

expression = "4*eps*((sig/r)^12-(sig/r)^6); "
expression += "eps=epsilon(type1, type2); "
expression += "sig=sigma(type1, type2)"

lj = mm.CustomNonbondedForce(expression)
lj.addTabulatedFunction("epsilon", mm.Discrete2DFunction(numTypes, numTypes, epsLST_r))
lj.addTabulatedFunction("sigma", mm.Discrete2DFunction(numTypes, numTypes, sigLST_r))
lj.addPerParticleParameter('type')
lj.setNonbondedMethod(mm.NonbondedForce.CutoffPeriodic)
lj.setCutoffDistance(0.9*unit.nanometer)
lj.setSwitchingDistance(0.6*unit.nanometer)
lj.setUseSwitchingFunction(True)

lj.addParticle([0])
lj.addParticle([1])
########

#FORCE 2
ff_buck = []
ff_buck.append({"type1" : "X" , "type2" : "Y" , "A" : "0.02" , "rho" : "2" , "C6" : "  0.0    " })
Aparam_r = np.zeros([numTypes,numTypes])
Rho_r = np.ones([numTypes,numTypes])
C6_r = np.zeros([numTypes,numTypes])
for f in ff_buck:
    i = atomTypes.index(f["type1"])
    j = atomTypes.index(f["type2"])
    Aparam_r[i,j] = float(f["A"])
    Rho_r[i,j]    = float(f["rho"])
    C6_r[i,j]     = float(f["C6"])
    Aparam_r[j,i] = Aparam_r[i,j]
    Rho_r[j,i]    = Rho_r[i,j]
    C6_r[j,i]     = C6_r[i,j]
AparamLST_r = (Aparam_r).ravel().tolist()
RhoLST_r    = (Rho_r).ravel().tolist()
C6LST_r     = (C6_r).ravel().tolist()

expression = "A*exp(-r/B) - C/r^6; "
expression += "A=A(type1, type2); "
expression += "B=B(type1, type2); "
expression += "C=C(type1, type2)"
buck = mm.CustomNonbondedForce(expression)
buck.setNonbondedMethod(mm.NonbondedForce.CutoffPeriodic)
buck.addTabulatedFunction(
    'A', mm.Discrete2DFunction(numTypes, numTypes, AparamLST_r))
buck.addTabulatedFunction(
    'B', mm.Discrete2DFunction(numTypes, numTypes, RhoLST_r))
buck.addTabulatedFunction(
    'C', mm.Discrete2DFunction(numTypes, numTypes, C6LST_r))
buck.addPerParticleParameter('type')
buck.setCutoffDistance(0.9*unit.nanometer)
buck.setSwitchingDistance(0.6*unit.nanometer)
buck.setUseSwitchingFunction(True)

buck.addParticle([0])
buck.addParticle([1])
########

#electrostatics
pme = mm.NonbondedForce()
pme.setNonbondedMethod(4)
pme.setCutoffDistance(0.9*unit.nanometer)
pme.setEwaldErrorTolerance(1e-5)

pme.addParticle(0.5,0,0)
pme.addParticle(-0.5,0,0)

#build system
positions = [
    mm.Vec3(x=0, y=0, z=0),
    mm.Vec3(x=.5, y=0, z=0),
] * unit.nanometer
system = mm.System()
for i in range(len(positions)):
    system.addParticle(20)

### adding both of these with pme breaks it ###
lj.addInteractionGroup(set([0,1]),set([0,1]))
buck.addInteractionGroup(set([0,1]),set([0,1]))
system.addForce(pme)
platform = mm.Platform.getPlatformByName('CUDA')
###############################################

system.addForce(lj)
system.addForce(buck)

integrator = mm.LangevinIntegrator(300*unit.kelvin,0.1/unit.picosecond,1*unit.femtosecond)
context = mm.Context(system, integrator, platform)
context.setPositions(positions)
print(context.getState(getEnergy=True).getPotentialEnergy())
```

**Comment by peastman** [comment]
Thanks for the report!  I have a good guess of what the problem is.  I'll get the fix in soon.

**Comment by peastman** [comment]
The fix is in #3729.
